File: tracsmobile/src/main/java/edu/txstate/mobile/tracs/util/SettingsStore.java
package edu.txstate.mobile.tracs.util;

import android.content.Context;
import android.util.Log;

import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import edu.txstate.mobile.tracs.AnalyticsApplication;
import edu.txstate.mobile.tracs.R;
import edu.txstate.mobile.tracs.notifications.NotificationTypes;
import edu.txstate.mobile.tracs.util.http.HttpQueue;
import edu.txstate.mobile.tracs.util.http.requests.SettingsRequest;

public class SettingsStore {
    private static final String TAG = "SettingsStore";
    private static final boolean SETTING_ENABLED = true;
    private static final boolean SETTING_DISABLED = false;
    private static final String[] SETTINGS_NOT_IMPLEMENTED = {
            NotificationTypes.ASSESSMENT,
            NotificationTypes.GRADE,
            NotificationTypes.DISCUSSION,
            NotificationTypes.ASSIGNMENT
    };
    private JsonObject settings = new JsonObject();

    private static SettingsStore notificationSettings;

    private SettingsStore() {}

    public static SettingsStore getInstance() {
        if (notificationSettings == null) {
            notificationSettings = new SettingsStore(new JsonObject());
            notificationSettings.getSettingsFromStorage();
            notificationSettings.setDefaultSettings();
        }
        return notificationSettings;
    }

    private SettingsStore(JsonObject settings) {
        this.settings = settings;
    }

    public void put(String settingId, boolean isEnabled) {
        this.settings.addProperty(settingId, isEnabled);
        AppStorage.put(AppStorage.SETTINGS, this.settings.toString(), AnalyticsApplication.getContext());
    }

    public Boolean get(String settingId) {
        if (this.settings.has(settingId)) {
            return this.settings.get(settingId).getAsBoolean();
        }
        return SETTING_ENABLED;
    }

    private void setDefaultSettings() {
        for (String setting : SETTINGS_NOT_IMPLEMENTED) {
            put(setting, SETTING_DISABLED);
        }
    }

    private void getSettingsFromStorage() {
        String settings = AppStorage.get(AppStorage.SETTINGS, AnalyticsApplication.getContext());
        if (settings == null || "".equals(settings)) {
            return;
        }
        JsonParser parser = new JsonParser();
        this.settings = parser.parse(settings).getAsJsonObject();
    }

    public void saveSettings() {
        Context context = AnalyticsApplication.getContext();
        AppStorage.put(AppStorage.SETTINGS, SettingsStore.getInstance().toString(), AnalyticsApplication.getContext());
        String settingsUrl = context.getString(R.string.dispatch_base) +
                context.getString(R.string.dispatch_settings);

        HttpQueue.getInstance(AnalyticsApplication.getContext()).addToRequestQueue(
                new SettingsRequest(settingsUrl, response ->
                        Log.i(TAG, "Settings saved")
                ), null);
    }

    public JsonObject getSettings() {
        return this.settings;
    }

    public void clear() {
        notificationSettings = new SettingsStore(new JsonObject());
    }

    public String toString() {
        return this.settings.toString();
    }
}

File: tracsmobile/src/main/java/edu/txstate/mobile/tracs/util/SettingsStore.kt
package edu.txstate.mobile.tracs.util

import android.util.Log
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import edu.txstate.mobile.tracs.AnalyticsApplication
import edu.txstate.mobile.tracs.R
import edu.txstate.mobile.tracs.notifications.NotificationTypes
import edu.txstate.mobile.tracs.util.http.HttpQueue
import edu.txstate.mobile.tracs.util.http.requests.SettingsRequest

class SettingsStore private constructor() {
    companion object {
        private val TAG = "SettingsStore"
        private val SETTING_ENABLED = true
        private val SETTING_DISABLED = false
        private val SETTINGS_NOT_IMPLEMENTED = arrayOf(
                NotificationTypes.ASSESSMENT,
                NotificationTypes.GRADE,
                NotificationTypes.DISCUSSION,
                NotificationTypes.ASSIGNMENT
        )
        var instance: SettingsStore = SettingsStore()
            @JvmStatic
            get() {
                field.getSettingsFromStorage()
                field.setDefaultSettings()
                return field
            }
    }

    private var settings = JsonObject()

    constructor(settings: JsonObject): this() {
        this.settings = settings
    }

    fun put(settingId: String, isEnabled: Boolean) {
        this.settings.addProperty(settingId, isEnabled)
        AppStorage.put(AppStorage.SETTINGS, this.settings.toString(), AnalyticsApplication.getContext())
    }

    fun get(settingId: String): Boolean {
        if (this.settings.has(settingId)) {
            return this.settings.get(settingId).asBoolean
        }
        return SETTING_ENABLED
    }

    private fun setDefaultSettings() {
        SETTINGS_NOT_IMPLEMENTED.map { it -> put(it, SETTING_DISABLED) }
    }

    private fun getSettingsFromStorage() {
        val settings: String? = AppStorage.get(AppStorage.SETTINGS, AnalyticsApplication.getContext())
        if (settings.isNullOrEmpty()) {
            return
        }
        val parser = JsonParser()
        this.settings = parser.parse(settings).asJsonObject
    }

    fun saveSettings() {
        val context = AnalyticsApplication.getContext()
        AppStorage.put(AppStorage.SETTINGS, SettingsStore.instance.toString(), context)
        val settingsUrl = context.resources.getString(R.string.dispatch_base) +
                context.resources.getString(R.string.dispatch_settings)
        HttpQueue.getInstance(context).addToRequestQueue(
                SettingsRequest(settingsUrl, { _ -> Log.i(TAG, "Settings saved") }),
                null)
    }

    fun getSettings(): JsonObject {
        return this.settings
    }

    fun clear() {
        instance = SettingsStore(JsonObject())
    }

    override fun toString(): String {
        return this.settings.toString()
    }
}